### ISSUE TITLE
CAL-AMP57 Ensure custom controls are keyboard accessible

### ISSUE THREAD

**Issue by dqateam** [Esri/calcite-components#14250]
**monday.com sync:** #11737229218

### Violation: 
 Ensure custom controls are keyboard accessible 

### WCAG Reference: 
- [2.1.1 Keyboard](https://www.w3.org/WAI/WCAG22/Understanding/keyboard.html) 

### Severity: 
 Critical 

### Media Type: 
 Custom Controls 

--- 
### Areas for Remediation: 
- Title: CAL-AMP57 Ensure custom controls are keyboard accessible 
- Module: M7 Table and color picker 
- Finding ID: [CAL-AMP57](https://esri.hub.essentia11y.com/short-link/YaWNlw1yQIEDtTbH) 

**Issue**
There are interactive "Color" buttons (present in the saved section of the color picker) that cannot be operated with the keyboard alone while a screen reader is running.

**User Impact**
Screen reader users will be prevented from accessing the functionality provided by these controls.

**Code Reference** 
 ```
<calcite-color-picker-swatch tabindex="0" class="saved-color" scale="m" calcite-hydrated=""></calcite-color-picker-swatch>

```

**Steps to Reproduce**
1. Enable a screen reader.
2. Navigate to the control using the Down Arrow key or the Tab key
3. Attempt to interact with Enter, Space, and arrow keys.
4. Notice that the control cannot be interacted with using the keyboard while the screen reader is running.
 
**Recommendation**
Ensure custom controls are keyboard accessible. Users must be able to navigate to and operate controls while a screen reader is running.

**Compliant Code Example**
```
<button aria-haspopup="true" aria-controls="menu1">Respond</button>
 <ul id="menu1" role="menu"
 tabindex="-1"
 aria-activedescendant="m1"
 aria-label="Respond with Emoji">
 <li id="m1" role="menuitem">
 Smile
 </li>
 <li id="m2" role="menu">
 Raise hand
 </li>
 <li id=m3" role="menu">
 Thumbs down
 </li>
 <li id=m4" role="menu">
 Thumbs up
 </li>
 </ul>

 <a id="join" href="#" role="button">Join</a>

<script>
 <!-- role button that activates on space and enter -->
 function join() { /* ... */ }
 document.getElementById("menu1").addEventListener ("keyup", event => { });
 document.getElementById("join").addEventListener ("keyup", event => { 
 if (event.keyCode === 32) {
 join();
 }
 });
</script>
```

 **Issue Image** 
 https://desktopqa/aic_lap/Calcite/Calcite2025/CAL-AMP57.png 
 

 --- 
### Additional Resources: 
- [LA Platform Report Link](https://esri.hub.essentia11y.com/short-link/YaWNlw1yQIEDtTbH) 
- [Esri Accessibility Resources](https://esriis.sharepoint.com/sites/Accessibility)
- Report Source: Product accessibility evaluation conducted on Calcite by Level Access and delivered August, 2025.

**Comment by geospatialem** [comment]
This was a follow-up issue created after consultation with Level Access discussions.

The reason this was imported is a result of the following discussion:

>>> ** [Kitty] CAL–AMP57 – Ensure custom controls are keyboard accessible**: I am able to keyboard navigate into the saved color. If I'm understanding the issue, this issue is requesting keyboard navigation to the saved color, is that correct? If its related to specific functionality the saved color is providing the context of the color previously selected, there is no specific functionality associated with selection. If its confusing, is the recommendation to not allow keyboard navigation to the saved color?
> 
>> **[a11y team]** Your understanding is mostly correct. After saving a color in the color picker, the issue occurs when a screen reader (NVDA) is enabled: activating a saved color using Enter or Space, it doesn't activate it. If a saved color is not intended to be interactive, it should not be keyboard focusable. However, if it is meant to be selectable, it must be implemented as an accessible control with proper keyboard and screen reader support. 
>
> [Kitty] It sounds like the recommendation would be to disable the keyboard navigation to the saved colors. I believe in the past this was disabled and flagged by Level Access since there were colors listed but not keyboard navigable though... so the team will need to consider if we want to implement the issue or leave it as-is for next steps.

**Comment by macandcheese** [comment]
Report mentions `color-picker-swatch` - the Color Picker now uses `swatch-group` and `swatch` so the current functionality may not match the tested behavior.

**Comment by geospatialem** [comment]
> Report mentions `color-picker-swatch` - the Color Picker now uses `swatch-group` and `swatch` so the current functionality may not match the tested behavior.

The behavior is the same with and without the `swatch`/`switch-group` components, however _believe_ the above is something that a developer would handle so am closing out the above as unplanned.

For instance: https://codepen.io/geospatialem/pen/raMgwOY, where the selected color change would be manipulated by the developer to change an input with a value, via either the picker selector, or when a saved color is selected using the component's `calciteColorPickerChange` event.

Marking the above as unplanned.
